samza-api/src/main/java/org/apache/samza/context/ApplicationDefinedTaskContext.java

```diff
@@ -0,0 +1,21 @@
+package org.apache.samza.context;
+
+/**
+ * An application should implement this to contain any runtime objects required to initialize the processing logic.
+ * <p>
+ * An instance of this will be created for each task.
+ * <p>
+ * If it is necessary to have a separate instance per task, then use {@link ApplicationDefinedContainerContext} instead.
+ * An {@link ApplicationDefinedTaskContextFactory} needs to be implemented to build instances of this context.
+ */
+public interface ApplicationDefinedTaskContext {
+  /**
+   * Initialization which will run before processing begins.
+   */
+  void start();
+
+  /**
+   * Shutdown logic which will run after processing ends.
+   */
+  void stop();
```

Comment:
1. Do we want to provide the configuration, metricsRegistry objects here to the user implementation or how will they get it?
2. We expect users to setup all the custom objects in the start implementation? Do the start/stop methods have to be idempotent by the high-level API? Would be better to document the contract here.
3. If the users have to use offspring factories internally here, are we going to offer any other alternative  other than static helper methods?